### PR TITLE
Fix for issue #3963

### LINKED ISSUE #3963: APPLY_FIXES and for PHP_PHPCSFIXER linter
**Describe the bug**
Violations found by PHP CS Fixer are not fixed.

**To Reproduce**

1. Create a `.mega-linter.yml` with the following content
```yaml
APPLY_FIXES: all
ENABLE_LINTERS:
  - PHP_PHPCSFIXER
```
2. Run MegaLinter locally with `npx mega-linter-runner --release v8.0.0` (I also tried `npx mega-linter-runner --release v8.0.0 --fix`)
3. See the output (which is not fixed)
```
Using cache file ".php-cs-fixer.cache".
   1) htdocs/packages/my-notifications/Classes/Notification/Notification.php (single_space_around_construct)
   2) htdocs/packages/my-crm/Classes/MyApi/Model/Lead/Lead.php (single_space_around_construct)
   3) htdocs/packages/my-crm/Classes/MyApi/Model/Lead/ContactDetails.php (single_space_around_construct)
   4) htdocs/packages/my-crm/Classes/MyApi/Model/Lead/LeadFieldsForCase.php (single_space_around_construct)
   5) htdocs/packages/my-crm/Classes/MyApi/Model/Lead/LeadSourceDetails.php (single_space_around_construct)
   6) htdocs/packages/my-crm/Classes/Utility/RegistryUtility.php (list_syntax)

Found 6 of 172 files that can be fixed in 0.551 seconds, 18.00 MB memory used
```

**Expected behavior**
PHP CS Fixer is called with `fix` (`php php-cs-fixer.phar fix ...`) and fixes the violations

### PR DESCRIPTION
**Pull request by llaville**
Fixes #3963

### Author/Contributor
- [ ] Add entry to the [CHANGELOG](https://github.com/oxsecurity/megalinter/blob/main/CHANGELOG.md) listing the change and linking to the corresponding issue (if appropriate)
- [ ] If documentation is needed for this change, has that been included in this pull request

### Reviewing Maintainer
- [ ] Label as `breaking` if this is a large fundamental change
- [ ] Label as either `automation`, `bug`, `documentation`, `enhancement`, `infrastructure`, or `performance`

### CODE REVIEW

**Comment by echoix** [comment]
I'll do a rerun of the failed ones. They aren't always 100% non-flaky..

**Comment by llaville** [comment]
> I'll do a rerun of the failed ones. They aren't always 100% non-flaky..

I've found origin of regression (introduced in my first commit, and fixed by second commit). Wait and see if CI is agree with my analysis, but locally it seems to be good ! 

Cross fingers ;-)

**Comment by llaville** [comment]
All seems ok except for saleforces_* linters (trivy scanner found vulnerabilities)

**Comment by echoix** [comment]
> All seems ok except for saleforces_* linters (trivy scanner found vulnerabilities)

They also appear in the auto update PR since yesterday

**Comment by echoix** [comment]
> All seems ok except for saleforces_* linters (trivy scanner found vulnerabilities)

Do you want it merged now ? If so, when ready, change the title to remove the Work in progress, and I'll do a final read and merge

**Comment by llaville** [comment]
> > All seems ok except for saleforces_* linters (trivy scanner found vulnerabilities)
> 
> Do you want it merged now ? If so, when ready, change the title to remove the Work in progress, and I'll do a final read and merge

I need to write some doc and changelog. I will do in next hours.

**Comment by llaville** [comment]
Finally found another problem. Need to investigate more. I'll be busy tomorrow, so I'll do it only thursday.

**Comment by llaville** [comment]
Finally, found it ! Now it's time for explains.

On my first approach and commit, I've introduced a test 
https://github.com/oxsecurity/megalinter/pull/3974/commits/fb1510a841bc1ac1a65dc9a70b92411e0c307625#diff-6edf526a9f6b62ef500e048620fd0d43bbc65957ed84e9033d2ff9f8f03206b4R259

that was usefull for PHPCSFIXER but add a regression for other linters.

So, I go back with second commit 
https://github.com/oxsecurity/megalinter/pull/3974/commits/0661847cae293c2db45aa08953828f04f7a6b075

It was an error, because even if it restored previous good behaviour for other linters, it won't work anymore for PHPCSFIXER.


My final version that works for all linters is :

- extracted code that activate apply_fixes feature (more pretty and easy to debug) on new function `def manage_apply_fixes(self, params)`

That gave an extra debug log for each linter; for example : 

```
[Activation] + PHP_PHPCSFIXER (PHP) was activated by ENABLE strategies
[Apply Fixes] is enabled for + PHP_PHPCSFIXER (PHP)
```

- define default apply_fixes value to False and remove first if test condition that was source or problem 

`if self.cli_lint_fix_arg_name is None:`

- restore fix test file status to original value (before apply fixes), that was in wrong state
https://github.com/oxsecurity/megalinter/pull/3974/commits/fb1510a841bc1ac1a65dc9a70b92411e0c307625#diff-6a17d64fcb7bf9898f3d9d0f22cd88bcd6bd083ce6dab6abc74ffaabcb106f4e


Last but not least, I've also :

- added note on PHPCSFIXER about apply_fixes
- added entry into CHANGELOG

That's all !

**Comment by nvuillam** [comment]
@llaville amazing, thanks for your hard work :)
I'll chech your PR asap :)

File: docs/descriptors/php_php_cs_fixer.md [hunk 1]
```diff
@@ -20,6 +20,13 @@ description: How to use php-cs-fixer (configure, ignore files, ignore errors, he
 - Enable php-cs-fixer by adding `PHP_PHPCSFIXER` in [ENABLE_LINTERS variable](https://megalinter.io/beta/configuration/#activation-and-deactivation)
 - Disable php-cs-fixer by adding `PHP_PHPCSFIXER` in [DISABLE_LINTERS variable](https://megalinter.io/beta/configuration/#activation-and-deactivation)
 
+- Enable **autofixes** by adding `PHP_PHPCSFIXER` in [APPLY_FIXES variable](https://megalinter.io/beta/configuration/#apply-fixes)
+
+> [!NOTE]
```
Comment (nvuillam) [inline docs/descriptors/php_php_cs_fixer.md:25]: @llaville this markdown file is generated, so if you need such text here, you need to define it in the descriptor YML file , in linter_text property :)

File: megalinter/Linter.py [hunk 2]
```diff
@@ -1260,12 +1273,14 @@ def build_lint_command(self, file=None) -> list:
         # Add fix argument if defined
         if self.apply_fixes is True and (
             self.cli_lint_fix_arg_name is not None
+            or len(self.cli_lint_fix_remove_args) > 0
```
Comment (nvuillam) [inline megalinter/Linter.py:1276]: @llaville that's too vague, I need to understand what you want to do here 🥺

**Comment by nvuillam** [comment]
@llaville I just merged, please can you make another PR to fix the doc in linter_text descriptor property ? :)

**Comment by llaville** [comment]
@nvuillam new PR #4249 to fix the doc

**Pull request merged**